**Step 2:**

## Cleaning and Transforming Data ##

Whenever we get data from any source, be it a csv file, sql or sourced from any place else, it would most often be scattered, raw and in such a form that no analysis can be done on it, forget dashboarding, and that means that data would not be of any use to us, which is why after loading the data, we always have to ensure that we clean and transform the data to extract useful information from it, that can help us get insight into the reason of customers churning.

In [5]:
import pandas as pd

raw_data = pd.read_csv(r'C:\Users\Kshitij Sharma\Customer Churn Project/Data/Raw_Data/Telco_Customer_Churn.csv')

raw_data.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

As we can see in our dataframe, we have column "TotalCharges" as 'object', which can cause issues when visualizing this data since Total charges indicates a count of something, and we usually count numbers, so we can put Total charges as an integer using **.to_numeric()** function from the pandas library.

In [6]:
raw_data['TotalCharges'] = pd.to_numeric(raw_data['TotalCharges'])

ValueError: Unable to parse string " " at position 488

While converting the "TotalCharges" column's data type from object to int, we received a **ValueError** indicating that there is a string stored at row 488 of this column, which means that there is a " " in row 488 of the column.


In [8]:
raw_data['TotalCharges'] = pd.to_numeric(raw_data['TotalCharges'], errors= 'coerce')

As we can see now, **.to_numeric()** function was accepted and it converted the 'TotalCharges' column's datatype into an integer. 

Depending on the context, we can always use an argument **(errors = 'coerce')** that allows us to coerce invalid parsable entries like " " or "" and convert them into NaN, avoiding the error altogether.

In [11]:
print(raw_data['Churn'].unique())

['No' 'Yes']


**.unique()** function from the pandas library displays the unique values(not multiples or duplicates) that are present in the column.

In [12]:
raw_data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

As we can observe our data, we have "Churn" column that has only two values 'yes' and 'no'. So, we can convert this to a numeric column that with 0 representing 'yes' and 1 representing 'No'.

To help us achieve that, we can use **map()** function from the pandas library that will help us map "yes" and "no" to 0 and 1 as dictionary's key-value pairs.

In [13]:
print(raw_data)

      customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1        

In [15]:
raw_data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

In [16]:
churn_values = {"Yes": 0, "No": 1}

raw_data['Churn'] = raw_data['Churn'].map(churn_values)

print(raw_data)

      customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1        

**IMPORTANT** 

I encountered an error, when i mapped "yes" to 1, and "no" to 0. It caused the program to display the values of each row of column "Churn" as NaN. 

To fix it, i went through the code, and ran it step by step from the beginning, running each line of code, and it worked!

As we have now converted the columns into data types that are more suitable for the columns we have in our dataframe, like the categorical ones get assigned text data types and the ones that are numerical get assigned a numerical column, so that it becomes easier to perfrom calculations and analysis on our data.

Since out data is now looking good with no missing values, and the ones we had, we took care of them. Since we now have clean data, we can now do feature engineering on it, and that will help us identify what features matter to gain insights that help us prevent a higher churn rate.


In [17]:
raw_data.to_csv(r'Customer_churn_clean_data.csv', index=False)

**to_csv()** function from pandas library allows us to be able to export the data in the form of a csv file. 